api/mapping/services_nlp.py

```diff
@@ -1,19 +1,305 @@
+from data.models import Concept, ConceptRelationship
+from django.db.models import Q
 
+from .models import (DataDictionary, NLPModel, ScanReport, ScanReportAssertion,
+                     ScanReportConcept, ScanReportField, ScanReportValue)
+from .services import find_standard_concept, get_concept_from_concept_code
 
-def nlp_request(search_term):
-    return True
 
+def get_data_from_nlp(url, headers, post_response_url):
+    """
+    This function takes a list of POST'ed URLs and returns
+    a list of responses.
+    """
 
+    get_response = []
+    for url in post_response_url:
 
-def nlp_single_string(pk, dict_string):
+        req = requests.get(url, headers=headers)
+        job = req.json()
+
+        while job["status"] != "succeeded":
+            req = requests.get(url, headers=headers)
+            job = req.json()
+            print("Waiting...")
+            time.sleep(3)
+        else:
+            get_response.append(job["results"])
+            print("Done!")
+
+    return get_response
+
+
+def process_nlp_response(get_response):
+    """
+    This function takes as input an NLP GET response
+    and returns a list of concept codes.
+
+    Input: get_response - A list of GET responses
+    Output: codes - A list of concept Codes
+
+    """
+    keep = ["ICD9CM", "ICD10CM", "RXNORM", "SNOMEDCT_US", "SNOMED"]
+    codes = []
+
+    for url in get_response:
+        for dict_entry in url["documents"]:
+            for entity in dict_entry["entities"]:
+                if "links" in entity.keys():
+                    for link in entity["links"]:
+                        if link["dataSource"] in keep:
+                            codes.append(
+                                [
+                                    dict_entry["id"],
+                                    entity["text"],
+                                    entity["category"],
+                                    entity["confidenceScore"],
+                                    link["dataSource"],
+                                    link["id"],
+                                ]
+                            )
+
+    print('CODES >>>', codes)
+    return codes
+
+
+def concept_code_to_id(codes):
+    """
+    This functions looks up standard and valid conceptIDs for concept codes
+    Appends conceptID to item in list, returns a dictionary of data
+    """
+    codes_dict = []
+    keys = [
+        "pk",
+        "nlp_entity",
+        "nlp_entity_type",
+        "nlp_confidence",
+        "nlp_vocab",
+        "nlp_code",
+        "conceptid",
+    ]
+    for item in codes:
+        x = get_concept_from_concept_code(
+            concept_code=str(item[5]), vocabulary_id=str(item[4])
+        )
+        item.append(x[1].concept_id)
+        codes_dict.append(dict(zip(keys, item)))
+
+    return codes_dict
+
+
+def start_nlp(search_term):
+
+    print(">>>>> Running NLP in services_nlp.py for", search_term)
+    field = ScanReportField.objects.get(pk=search_term)
```

Comment:
What is search_term? Is it the scanreportfield id?

Reply:
Yes. Do you want me to change this to something else @py5gol?